lib/_stream_readable.js

```diff
@@ -1070,13 +1072,90 @@ Readable.prototype.wrap = function(stream) {
 };
 
 Readable.prototype[SymbolAsyncIterator] = function() {
-  if (createReadableStreamAsyncIterator === undefined) {
-    createReadableStreamAsyncIterator =
-      require('internal/streams/async_iterator');
+  let stream = this;
+
+  if (typeof stream.read !== 'function') {
+    // v1 stream
+    const src = stream;
+    stream = new Readable({
+      objectMode: true,
+      destroy(err, callback) {
+        destroyImpl.destroyer(src, err);
+        callback();
+      }
+    }).wrap(src);
   }
-  return createReadableStreamAsyncIterator(this);
+
+  const iter = createAsyncIterator(stream);
+  iter.stream = stream;
+  return iter;
 };
 
+async function* createAsyncIterator(stream) {
+  const state = stream._readableState;
+
+  if ((state && state.endEmitted) ||
+      (state && state.closeEmitted)) {
+    // TODO (ronag): What if errorEmitted?
+    // TODO (ronag): Wait for closeEmitted?
+    // TODO (ronag): if (closeEmitted && !endEmitted) ERR_PREMATURE_CLOSE?
+    stream.on('error', nop);
+    destroyImpl.destroyer(stream, null);
+    return;
+  }
+
+  let callback = nop;
+  let error;
+  let ended = false;
```

Comment:
I am wondering if it's not sufficient to only resolve the callback on all these events and check `stream._readableState` for the truth value instead of having these 3 extra variables.

Reply:
I was trying to maximize compatibility in case users try to re-use `Readable.prototype[Symbol.asyncIterator]` on their stream like objects. Maybe a bit too paranoid. Not sure where this is in relation to `emit('error', undefined)` :D.

Reply:
> I was trying to maximize compatibility in case users try to re-use Readable.prototype[Symbol.asyncIterator] on their stream like objects.

I think that wouldn't work anyway since we [already read _readableState](https://github.com/nodejs/node/pull/34035/files#diff-ba6a0df0f5212f5cba5ca5179e209a17R1095)

> Not sure where this is in relation to emit('error', undefined)

Someone can create a regular readable stream and call `.emit('error')` without an argument to signal an error. I am not saying it's a good idea but it doesn't require any custom code since we expose `.emit` on readable streams.

Reply:
> I think that wouldn't work anyway since we already read _readableState

Why? It's only optionally used.

> Someone can create a regular readable stream and call

It's fine. I was just joking.

Reply:
> Why? It's only optionally used.

Optionally used? If a user passes their own ended stream (with Node stream like semantics) and the stream has no _readableState than this implementation would be incorrect for it (right?).

I am fine with not having any pretense of this method working on anything other than readable streams.

Reply:
Fixed.

Reply:
Reverted. We need to have state for `'error'` anyway to handle users manually emitting `'error'`.